pkg/middleware/auth_proxy_test.go

```diff
@@ -0,0 +1,89 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/remotecache"
+	authproxy "github.com/grafana/grafana/pkg/middleware/auth_proxy"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/require"
+	macaron "gopkg.in/macaron.v1"
+)
+
+// Test initContextWithAuthProxy with a cached user ID that is no longer valid.
+//
+// In this case, the cache entry should be ignored/cleared and another attempt should be done to sign the user
+// in without cache.
+func TestInitContextWithAuthProxy_CachedInvalidUserID(t *testing.T) {
+	const name = "markelog"
+	const userID = int64(1)
+	const orgID = int64(4)
+
+	upsertHandler := func(cmd *models.UpsertUserCommand) error {
+		require.Equal(t, name, cmd.ExternalUser.Login)
+		cmd.Result = &models.User{Id: userID}
+		return nil
+	}
+	getSignedUserHandler := func(cmd *models.GetSignedInUserQuery) error {
+		// Simulate that the cached user ID is stale
+		if cmd.UserId != userID {
+			return models.ErrUserNotFound
+		}
+
+		cmd.Result = &models.SignedInUser{
+			UserId: userID,
+			OrgId:  orgID,
+		}
+		return nil
+	}
+
+	origHeaderName := setting.AuthProxyHeaderName
+	origEnabled := setting.AuthProxyEnabled
+	origHeaderProperty := setting.AuthProxyHeaderProperty
+	origUpsertHandler := bus.AddHandlerForType(&models.UpsertUserCommand{}, upsertHandler)
```

Comment:
Is the introduction/usage of `AddHandlerForType` really needed. We haven't really had a problem with this before - what kind of problem did you get?

Reply:
@marefr It's necessary for cleaning up handlers after the test finishes. I couldn't see any other way to restore the original handlers. I think the original API is flawed TBH, since it implicitly derives the handler key, and this explicit API is better.

Reply:
Agree with you, but sure it's the path we want to go. Moving forward the goal should be to remove the usage of the global bus to allow writing tests in separation without having to bother about state.

Reply:
@marefr, I've also seen that we plan to get rid of the global bus. But as it stands I don't see how to write this test differently?

Reply:
I tend to agree with @marefr 
Moreover because `origUpsertHandler` and `origSignedUserHandler` are `nil` which mean that my original assumption that these handlers were registered was wrong.
Anyway, even though I don't insist, I would prefer to address this everywhere instead of applying to a specific test.

Reply:
> I don't see how to write this test differently?

If you skip setting back the handlers does the tests fail? If not, you're all set and don't need the `AddHandlerForType`.

Reply:
@marefr As I explained before I think it's a very bad idea to not restore state changed by tests. Although it _might_ not affect anything now, it's an assumption that breaks very easily once you write more tests. Also, I think that objectively the new method is better, since it's explicit about how the key is derived (i.e. from the message type).

@bergquist @kylebrandt what do you think?